### ISSUE TITLE
Support for early termination

### ISSUE THREAD

**Issue by maoe** [Gabriel439/Haskell-Foldl-Library#85]
Is there a good way to support early termination in the foldl library? Currently `fold` has to go though the list until the end even if `head` is passed.

**Comment by Gabriel439** [comment]
No, this library does not support any way for the `Fold` to indicate that no further input is necessary

**Comment by michaelt** [comment]
There is an interesting systematic attempt at this in https://github.com/effectfully/prefolds One simple-minded point in defense of the `foldl` approach is that the emphasis of the library is on combining simultaneous folds e.g. `liftA3 (,,) L.head (L.index 17) L.sum` so though individual stock folds could exit early, their combinations tend to consume all of the input.

**Comment by maoe** [comment]
@michaelt Thanks for the info. That's interesting. I was thinking of a simpler solution like `x -> a -> Maybe x` or `x -> a -> (Bool, x)` for the step function in `Fold`. The argument about early termination with composition is a good point but I think it's still useful to support it.

I guess I'll see if I can support it without sacrificing the simple API and good performance.

**Comment by rizo** [comment]
@Gabriel439 could you please elaborate on the decision of not including early termination in Foldl? Are you aware of any structures similar to Foldable that support termination in reducing functions?

**Comment by Gabriel439** [comment]
@rizo: I'm not aware of a similar structure that supports (A) early termination and (B) `Applicative` composition of `Fold`s

The main reason why `foldl` does not do so is because the scope of the library is to be a replacement in spirit for `Data.List.foldl`.  That's why the signature of the `Fold` type exactly matches the arguments you pass to `foldl` (which also does not have a termination check).  There is a similar library named `folds` with a broader scope where this might be a better fit:

http://hackage.haskell.org/package/folds

**Comment by joelmccracken** [comment]
Is there *any* way to do this? I am trying to wrap a process that watches for changes on a value, and so intentionally does not terminate; as it stands, it appears that I have no way to use the abstractions provided by turtle to do this. 

Here is an example that demonstrates the problem:

```
*Main Lib Paths_etcdctl> import Turtle
*Main Lib Paths_etcdctl Turtle> import qualified Control.Foldl as Fold
*Main Lib Paths_etcdctl Turtle Fold> :t fold (inproc "yes" [] mempty) Fold.head
fold (inproc "yes" [] mempty) Fold.head
  :: MonadIO io => io (Maybe Line)
*Main Lib Paths_etcdctl Turtle Fold> fold (inproc "yes" [] mempty) Fold.head
```

which blocks indefinitely. Now I know you are saying in this issue that this wont make sense for the foldl library, but I also cannot seem to make it work with the lower-level function, e.g. 

```
*Main Lib Paths_etcdctl Turtle Control.Foldl Fold> foldShell (inproc "yes" [] mempty) (FoldShell (\_ a-> pure $ Just a) Nothing pure)
```

I saw this answer on stack overflow and it *seemed* to perhaps show a way: 

https://stackoverflow.com/questions/28421469/how-to-drop-lines-when-streaming-from-a-file-using-haskell-and-the-turtle-librar

But testing it:

```
*Main Lib Paths_etcdctl Turtle Control.Foldl Fold> view $ limit 10 (inproc "yes" [] mempty)
Line "y"
Line "y"
Line "y"
Line "y"
Line "y"
Line "y"
Line "y"
Line "y"
Line "y"
Line "y"

```
... which still blocks indefinitely.

**Comment by joelmccracken** [comment]
... I entered this in the wrong project. Sorry. Deleting these comment and adding elsewhere.
